### DISCUSSION TITLE
hinaOS上で動かせる実行ファイルを作りたい

### DISCUSSION THREAD

**Discussion by keisuke713** [nuta/microkernel-book#17] [Help] [answered]
### 概要

最終的にやりたいこととしてhinaOS上で下記のように実行ファイルを動作させたいと考えていて、そのためにまず実行ファイルを作りたいと考えています。
```
shell > ./a.out
=> do something
```

しかし、リンクで少し手こずっています。
コンパイル時に `-I/Users/nebashikeisuke/c/microkernel-book` などサーチ元を指定すればプリプロセス・コンパイルの段階ではエラーは出ませんが、リンク時にはエラーが出ると思います。
しかしリンク時のエラーを解消しようとファイルを追加で指定すればするほど、更に指定しなければいけないファイルが増えるという無限ループに陥っています。。
今回実行ファイルにしたいのは[こちら](https://github.com/keisuke713/microkernel-book/blob/fix/executable/servers/test/main.c)なのですが、このシンプルなコードをコンパイルするのにも以下のように大量のファイルを指定する必要がありました。(エントリーポイントを指定していないのでおそらくこれでも足りない)
```
nebashikeisuke@nebashiukunoAir microkernel-book % clang -I/Users/nebashikeisuke/c/microkernel-book -I/Users/nebashikeisuke/c/microkernel-book/libs/user/riscv32 servers/test/main.c libs/user/printf.c libs/common/string.c libs/common/vprintf.c libs/user/syscall.c libs/common/riscv32/backtrace.c libs/common/backtrace.c libs/common/symbol_table.S --target=riscv32 -g3 -std=c11 -ffreestanding -fno-builtin -nostdlib -nostdinc
```

macOSやlinuxでのリンクのように標準ライブラリは明示的に指定しなくても勝手にリンクしてくれれば良いのに。と思うのですが、何か良い方法はありますでしょうか？

### ソースコードのURL

https://github.com/keisuke713/microkernel-book/tree/fix/executable

**Comment by nuta** [comment, marked as answer]
`servers/hello` 配下に最小限のアプリケーションのサンプルコードがあります。こちらで遊んでみるのはどうでしょうか。

具体的なリンカーへの引数は、`make V=1`で確認できます:

```
        LD  build/servers/hello.elf
mkdir -p build/servers
/opt/homebrew//opt/llvm/bin/ld.lld  -Tbuild/servers/hello/user.ld -Map build/servers/hello.map -o build/servers/hello.elf.tmp build/servers/hello/main.o  build/libs/common.o  build/libs/user.o build/program_names/hello.o
```

> macOSやlinuxでのリンクのように標準ライブラリは明示的に指定しなくても勝手にリンクしてくれれば良いのに

なぜなら、ビルド時に何が起きているのか (どのようなファイルが、どのような設定でリンクされるのか）が隠れてしまいデバッグが困難になるためです。もちろんCコンパイラに手を加えれば一応実現できるのですが、それをするとHinaOSのビルドを理解するのにClangの内部実装の知識が必要になるため、意図的に避けています。

**Reply by keisuke713** [reply]
> servers/hello 配下に最小限のアプリケーションのサンプルコードがあります。こちらで遊んでみるのはどうでしょうか。

こちら正しく僕が作成したかったものです！こちらで遊びつつ最終的にshellから実行できるところまで持っていけたらなと思います。

> 具体的なリンカーへの引数は、make V=1で確認できます:
/opt/homebrew//opt/llvm/bin/ld.lld  -Tbuild/servers/hello/user.ld -Map build/servers/hello.map -o build/servers/hello.elf.tmp build/servers/hello/main.o  build/libs/common.o  build/libs/user.o build/program_names/hello.o

なるほど、アプリケーションの実行ファイルを作る際には `common.o` と `user.o` をマップさせているのですね。
一つずつ全てのファイルを指定するよりシンプルになりますね。

> ビルド時に何が起きているのか (どのようなファイルが、どのような設定でリンクされるのか）が隠れてしまいデバッグが困難になるためです。

clangの設定オプションで何かあれば良いなと考えていましたが、それだと確かに処理が隠蔽されてしまい逆に大変でしたね。

理解できました、いつも丁寧な回答ありがとうございます 🙏
